src/libstd/sys/unix/process/process_fuchsia.rs

```diff
@@ -0,0 +1,177 @@
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use io;
+use libc;
+use mem;
+use ptr;
+
+use sys::mx_cvt;
+use sys::magenta::{Handle, launchpad_t, mx_handle_t};
+use sys::process::process_common::*;
+
+////////////////////////////////////////////////////////////////////////////////
+// Command
+////////////////////////////////////////////////////////////////////////////////
+
+impl Command {
+    pub fn spawn(&mut self, default: Stdio, needs_stdin: bool)
+                 -> io::Result<(Process, StdioPipes)> {
+        if self.saw_nul() {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput,
+                                      "nul byte found in provided data"));
+        }
+
+        let (ours, theirs) = self.setup_io(default, needs_stdin)?;
+
+        let (launchpad, process_handle) = unsafe { self.do_exec(theirs)? };
+
+        Ok((Process { launchpad: launchpad, handle: Handle::new(process_handle) }, ours))
+    }
+
+    pub fn exec(&mut self, default: Stdio) -> io::Error {
+        if self.saw_nul() {
+            return io::Error::new(io::ErrorKind::InvalidInput,
+                                  "nul byte found in provided data")
+        }
+
+        match self.setup_io(default, true) {
+            Ok((_, _)) => {
+                // FIXME: This is tough because we don't support the exec syscalls
+                unimplemented!();
+            },
+            Err(e) => e,
+        }
+    }
+
+    unsafe fn do_exec(&mut self, stdio: ChildPipes)
+                      -> io::Result<(*mut launchpad_t, mx_handle_t)> {
+        use sys::magenta::*;
+
+        let job_handle = mxio_get_startup_handle(mx_hnd_info(MX_HND_TYPE_JOB, 0));
+        let envp = match *self.get_envp() {
+            Some(ref envp) => envp.as_ptr(),
+            None => ptr::null(),
+        };
+
+        // To make sure launchpad_destroy gets called on the launchpad if this function fails
+        struct LaunchpadDestructor(*mut launchpad_t);
+        impl Drop for LaunchpadDestructor {
+            fn drop(&mut self) { unsafe { launchpad_destroy(self.0); } }
+        }
+
+        let mut launchpad: *mut launchpad_t = ptr::null_mut();
+        let launchpad_destructor = LaunchpadDestructor(launchpad);
+
+        // Duplicate the job handle
+        let mut job_copy: mx_handle_t = MX_HANDLE_INVALID;
+        mx_cvt(mx_handle_duplicate(job_handle, MX_RIGHT_SAME_RIGHTS,
+                                   &mut job_copy as *mut mx_handle_t))?;
+        // Create a launchpad
+        mx_cvt(launchpad_create(job_copy, self.get_argv()[0],
+                                &mut launchpad as *mut *mut launchpad_t))?;
```

Comment:
`launchpad` is actually a regular `*mut launchpad_t`. It get's `Copy`ed as `LaunchpadDestructor.0`. Should "move" it into the `LaunchpadDestructor` and `s/LaunchpadDestructor/Launchpad` instead?